### DISCUSSION TITLE
값, 객체 데이터 불변 보장에 관한 질문

### DISCUSSION THREAD

**Discussion by fkdl0048** [fkdl0048/Mentoring-Project#102] [Q&A] [answered]
안녕하세요 멘토님

리팩터링 9장을 읽다가 궁금한 부분이 생겨서 질문드립니다.

객체 내부에서 필드로 객체를 지니고 있는 구조(구성)에서 해당 필드 객체의 불변을 보장하는 방법으로 책에서 해당 내부 객체의 불변이 보장되고 (생성자로만 할당, readonly등) 이후에 상위 클래스에서 세터로 접근할때 마다 새로운 객체로 대입하는 방법을 소개하는데

이 처럼 내부 필드 객체의 불변을 보장하기 위해선 위와 같은 방법이 최적일까요?

https://www.csharpstudy.com/Mistake/Article/17

여기에서 말하는 인터페이스를 사용하는 방법이나 DeepCopy로 처리하는게 가장 대중적인 방법일까요?

전에 말씀해주신 레코드도 사용하기 좋은 것 같습니다.

**Comment by jongfeel** [comment, marked as answer]
방법이야 어떻든 간에 불변 객체라는 걸 보장할 수 있으면 됩니다.

오히려 생성자 주입이 가장 보편적인 방식이고 그 생성자 파라미터가 인터페이스를 구현한 객체라면 더할 나위 없겠죠.
인터페이스는 메소드만 노출할 수 있으니까요.

C#에 record가 추가 되었으니까 활용하면 더 좋긴 하겠죠. 그러면 copy를 할 때 고민할 필요 없이 with 키워드를 사용하면 값 자체를 복사하니까 deep copy냐 shallow copy냐를 논할 필요도 없겠죠.
